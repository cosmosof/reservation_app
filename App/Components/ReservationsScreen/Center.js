// @flow

import * as React from 'react';
import {View} from 'react-native';

import styles from './Styles/CenterStyles';

type Props = { style?: Object, children?: React.Node };

export default function Center(props: Props) {
  return <View style={[styles.container, props.style]}>{props.children}</View>;
}
