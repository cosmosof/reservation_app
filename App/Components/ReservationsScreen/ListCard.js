// @flow

import * as React from 'react';
import {Text, View} from 'react-native';
import styles from './Styles/ListCardStyles';

type Props = { data: Object, styles: Object };

export default function ListCard(props: Props) {
  const {container, value, label, oddRow, row} = styles;
  const {name, id, hotelName, arrivalDate, departureDate} = props.data;
  return (
    <View
      key={id}
      style={container}
    >
      <View style={oddRow}>
        <Text style={label}>Guest name </Text>
        <Text style={value}> {name} </Text>
      </View>
      <View style={row}>
        <Text style={label}>Hotel name </Text>
        <Text style={value}> {hotelName} </Text>
      </View>
      <View style={oddRow}>
        <Text style={label}>Arrival date</Text>
        <Text style={value}> {arrivalDate} </Text>
      </View>
      <View style={row}>
        <Text style={label}>Departure date</Text>
        <Text style={value}> {departureDate} </Text>
      </View>
    </View>
  );
}
