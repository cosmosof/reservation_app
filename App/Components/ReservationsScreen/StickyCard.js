// @flow

import * as React from 'react';
import {View} from 'react-native';

import styles from './Styles/StickyCardStyles';

type Props = { children?: React.Node };

export default function StickyCard(props: Props) {
  return (
    <View style={styles.container}>
      {props.children}
    </View>
  );
}
