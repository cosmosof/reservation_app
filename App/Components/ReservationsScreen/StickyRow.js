// @flow

import * as React from 'react';
import {View} from 'react-native';

import styles from './Styles/StickyRowStyles';

type Props = { style?: Object, children?: React.Node };

export default function StickyRow(props: Props) {
  return <View style={[styles.container, props.style]}>{props.children}</View>;
}
