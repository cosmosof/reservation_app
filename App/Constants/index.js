import Colors from './Colors';
import Fonts from './Fonts';
import Metrics from './Metrics';
import Images from './Images';
import ListFilters from './ListFilters';
import ApplicationStyles from './ApplicationStyles';
import TestHelpers from './TestHelpers';

export {Colors, Fonts, Images, Metrics, ListFilters, ApplicationStyles, TestHelpers};
