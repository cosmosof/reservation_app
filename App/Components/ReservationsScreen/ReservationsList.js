// @flow

import React from 'react';
import {FlatList, Text, View} from 'react-native';
import {Query} from 'react-apollo';
import {RESERVATIONS} from '../../Api/Queries';
import ListCard from '../../Components/ReservationsScreen/ListCard';
import AlertMessage from '../../Components/Common/AlertMessage';

import styles from './Styles/ReservationsListStyles';

type Props = { filter: string, text: string };

export default function ReservationList(props: Props) {
  const {filter, text} = props;
  const {warningContainer, warningText} = styles;

  return (
    <Query
      query={RESERVATIONS}
      variables={{where: {[filter]: text}}}
    >
      {({loading, error, data, fetchMore, networkStatus, refetch}) => {
        const {reservations} = data;
        if (loading) {
          return (
            <View style={warningContainer}>
              <AlertMessage
                show
                title="loading..."
              />
            </View>
          );
        }
        if (error) {
          return (
            <View style={warningContainer}>
              <AlertMessage
                show
                title="An error occured"
              />
            </View>
          );
        }
        if (reservations && reservations.length === 0) {
          return (
            <View style={warningContainer}>
              <Text style={warningText}>Nothing found...</Text>
            </View>
          );
        }

        return (
          <FlatList
            data={reservations}
            keyExtractor={(item, index) => item.id}
            onEndReached={() =>
              fetchMore({
                variables: {
                  skip: reservations.length + 20,
                },
                updateQuery: (prev, {fetchMoreResult}) => {
                  if (!fetchMoreResult) return prev;
                  return Object.assign({}, prev, {
                    reservations: [...prev.reservations, ...fetchMoreResult.reservations],
                  });
                },
              })
            }
            onEndThreshold={0.5}
            onRefresh={() => refetch()}
            refreshing={networkStatus === 4}
            renderItem={({item}) => <ListCard data={item} />}
          />
        );
      }}
    </Query>
  );
}
