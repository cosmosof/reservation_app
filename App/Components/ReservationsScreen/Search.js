// @flow

import * as React from 'react';
import {TextInput, View, Platform} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import {Colors} from '../../Constants';

import styles from './Styles/SearchStyles';

type Props = {
  children: Function,
  styles: Object,
};

type State = {
  text: string,
};

export default class Search extends React.Component<Props, State> {
  constructor(props: Object) {
    super(props)
    this.state = {
      text: '',
    };
  }
  
  updateText = (text: string) => {
    this.setState({text});
  };
  render() {
    const {text} = this.state;
    return (
      <View style={styles.searchBarContainer}>
        <View style={styles.searchBar}>
          <Icon
            color={Colors.charcoal}
            name={Platform.OS == 'ios' ? 'ios-search' : 'md-search'}
            size={20}
          />
          <TextInput
            autoCorrect={false}
            onChangeText={(text) => this.updateText(text)}
            placeholder="search reservations..."
            style={styles.textInput}
            value={text}
          />
        </View>
        {this.props.children(text)}
      </View>
    );
  }
}
