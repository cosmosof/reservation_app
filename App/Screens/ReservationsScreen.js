import React from 'react';
import {FlatList, Picker, SafeAreaView} from 'react-native';
import {Query} from 'react-apollo';
import Icon from 'react-native-vector-icons/MaterialIcons';
import AlertMessage from '../Components/Common/AlertMessage';
import Button from '../Components/Common/Button';
import Center from '../Components/ReservationsScreen/Center';
import ListCard from '../Components/ReservationsScreen/ListCard';
import Label from '../Components/Common/Label';
import Search from '../Components/ReservationsScreen/Search';
import StickyCard from '../Components/ReservationsScreen/StickyCard';
import Row from '../Components/Common/Row';
import {ListFilters, Colors} from '../Constants';
import {RESERVATIONS} from '../Api/Queries';

import styles from './Styles/ReservationScreenStyles';

type Props = {};

type State = {
  filter: string,
  searchText: string,
};

export default class ReservationsScreen extends React.Component<Props, State> {
  static navigationOptions = {
    headerTitle: <Label
      textStyle={styles.navbarHeader}
      title="Reservations"
                 />,
  };
  constructor(props: Object) {
    super(props);
    this.state = {filter: 'name', searchText: ''};
  }
  handleFilter = (filter) => {
    this.setState({filter});
  };
  handleSearch = (searchText) => {
    this.setState({searchText});
  };
  render() {
    const {searchText, filter} = this.state;
    const {buttonStyle, container, filterLabel, itemStyle, pickerStyle, titleStyle, stickyFilter} = styles;
    const variables = searchText === '' ? {where: {}} : {where: {[filter]: searchText}};
    return (
      <SafeAreaView style={container}>
        <StickyCard>
          <Row>
            <Search>
              {(text) => (
                <Button
                  buttonStyle={buttonStyle}
                  onPress={() => this.handleSearch(text)}
                  textStyle={titleStyle}
                  title="Search"
                />
              )}
            </Search>
          </Row>
          <Row style={stickyFilter}>
            <Icon
              color={Colors.charcoal}
              name="filter-list"
              size={24}
            />
            <Label
              textStyle={filterLabel}
              title="search"
            />
            <Picker
              itemStyle={itemStyle}
              onValueChange={this.handleFilter}
              selectedValue={filter}
              style={pickerStyle}
            >
              {ListFilters.map(({index, name, value}) => (
                <Picker.Item
                  key={index}
                  label={name}
                  value={value}
                />
              ))}
            </Picker>
          </Row>
        </StickyCard>
        <Query
          query={RESERVATIONS}
          variables={variables}
        >
          {({loading, error, data, fetchMore, networkStatus, refetch}) => {
            const {reservations} = data;
            if (error) {
              return (
                <Center>
                  <AlertMessage
                    show
                    title="An error occured"
                  />
                </Center>
              );
            }
            if (loading) {
              return (
                <Center>
                  <AlertMessage
                    show
                    title="loading..."
                  />
                </Center>
              );
            }
            if (reservations && reservations.length === 0) {
              return (
                <Center>
                  <AlertMessage
                    show
                    title="Nothing found..."
                  />
                </Center>
              );
            }
            return (
              <FlatList
                data={reservations}
                keyExtractor={(item) => item.id}
                onEndReached={() =>
                  fetchMore({
                    variables: {
                      skip: reservations.length + 20,
                    },
                    updateQuery: (prev, {fetchMoreResult}) => {
                      if (!fetchMoreResult) return prev;
                      return Object.assign({}, prev, {
                        reservations: [...prev.reservations, ...fetchMoreResult.reservations],
                      });
                    },
                  })
                }
                onEndThreshold={0.5}
                onRefresh={() => refetch()}
                refreshing={networkStatus === 4}
                renderItem={({item}) => <ListCard data={item} />}
              />
            );
          }}
        </Query>
      </SafeAreaView>
    );
  }
}
