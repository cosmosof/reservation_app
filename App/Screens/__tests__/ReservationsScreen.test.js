import React from 'react';
import {shallow} from 'enzyme';
import ReservationsScreen from '../ReservationsScreen';
import {TestHelpers} from '../../Constants';

const {sampleText} = TestHelpers;

describe('<ReservationsScreen>', () => {
  const Component = <ReservationsScreen />;
  describe('Structure', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(Component);
    });
    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('renders a SafeAreaView', () => {
      expect(wrapper.find('SafeAreaView')).toHaveLength(1);
    });
    it('renders a StickyCard', () => {
      expect(wrapper.find('StickyCard')).toHaveLength(1);
    });
    it('renders two Rows', () => {
      expect(wrapper.find('Row')).toHaveLength(2);
    });
    it('renders a Search', () => {
      expect(wrapper.find('Search')).toHaveLength(1);
    });
    it('renders a Icon', () => {
      expect(wrapper.find('Icon')).toHaveLength(1);
    });
    it('renders a Search', () => {
      expect(wrapper.find('Search')).toHaveLength(1);
    });
    it('renders a Picker', () => {
      expect(wrapper.find('Picker')).toHaveLength(1);
    });
    it('renders a Label', () => {
      expect(wrapper.find('Label')).toHaveLength(1);
    });
    it('renders a Query', () => {
      expect(wrapper.find('Query')).toHaveLength(1);
    });
  });
  describe('Functionality', () => {
    const wrapper = shallow(Component);
    it('renders initial states correctly', () => {
      expect(wrapper.state('filter')).toBe('name');
      expect(wrapper.state('searchText')).toBe('');
    });
    it('sets states correctly', () => {
      wrapper.setState({filter: sampleText});
      wrapper.setState({searchText: sampleText});
      expect(wrapper.state('filter')).toEqual(sampleText);
      expect(wrapper.state('searchText')).toEqual(sampleText);
    });
    it('handleSearch sets state correctly', () => {
      wrapper.instance().handleSearch(sampleText);
      expect(wrapper.state('searchText')).toEqual(sampleText);
    });
    it('handleFilter sets state correctly', () => {
      wrapper.instance().handleFilter(sampleText);
      expect(wrapper.state('filter')).toEqual(sampleText);
    });
    it('handles onPress callback', () => {
      const render = wrapper.dive();
      render.find('Button').forEach((child) => {
        expect(child.simulate('press')).toHaveBeenCalledTimes(1);
      });
    });
  });
});
