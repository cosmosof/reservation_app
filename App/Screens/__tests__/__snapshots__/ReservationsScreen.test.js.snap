// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReservationsScreen> Structure renders correctly 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#F4F4F4",
      "flex": 1,
    }
  }
>
  <StickyCard>
    <Row>
      <Search>
        <Component />
      </Search>
    </Row>
    <Row
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      <Icon
        allowFontScaling={false}
        color="#595959"
        name="filter-list"
        size={24}
      />
      <Label
        textStyle={
          Object {
            "color": "#595959",
            "marginLeft": 10,
          }
        }
        title="search"
      />
      <Picker
        itemStyle={
          Object {
            "borderTopColor": "transparent",
            "color": "#595959",
            "fontFamily": "AvenirNext-Regular",
            "fontSize": 14,
            "height": 36,
          }
        }
        mode="dialog"
        onValueChange={[Function]}
        selectedValue="name"
        style={
          Object {
            "backgroundColor": "#F4F4F4",
            "borderRadius": 4,
            "marginLeft": 10,
            "width": 200,
          }
        }
      >
        <PickerItem
          key="0"
          label="guest name"
          value="name"
        />
        <PickerItem
          key="1"
          label="in guest names"
          value="name_contains"
        />
        <PickerItem
          key="3"
          label="name starts with"
          value="name_starts_with"
        />
        <PickerItem
          key="4"
          label="hotel name"
          value="hotelName"
        />
        <PickerItem
          key="5"
          label="in hotel names"
          value="hotelName_contains"
        />
        <PickerItem
          key="7"
          label="hotel names start with"
          value="hotelName_starts_with"
        />
        <PickerItem
          key="9"
          label="arrival date"
          value="arrivalDate"
        />
        <PickerItem
          key="10"
          label="departure date"
          value="departureDate"
        />
      </Picker>
    </Row>
  </StickyCard>
  <Query
    query={
      Object {
        "definitions": Array [
          Object {
            "directives": Array [],
            "kind": "OperationDefinition",
            "name": Object {
              "kind": "Name",
              "value": "Reservations",
            },
            "operation": "query",
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "where",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "where",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "first",
                      },
                      "value": Object {
                        "kind": "IntValue",
                        "value": "20",
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "skip",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "skip",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "reservations",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "hotelName",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "departureDate",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "arrivalDate",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
              ],
            },
            "variableDefinitions": Array [
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "ReservationWhereInput",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "where",
                  },
                },
              },
              Object {
                "defaultValue": undefined,
                "directives": Array [],
                "kind": "VariableDefinition",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "Int",
                  },
                },
                "variable": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "skip",
                  },
                },
              },
            ],
          },
        ],
        "kind": "Document",
        "loc": Object {
          "end": 209,
          "start": 0,
        },
      }
    }
    variables={
      Object {
        "where": Object {},
      }
    }
  >
    <Component />
  </Query>
</SafeAreaView>
`;
